### PR TITLE
Relax the validation of master ipv4 cidr for GKE with private endpoint subnetwork

### PR DESCRIPTION
**Pull request by ncapps**
Master ipv4 cidr is not required when a private endpoint subnetwork is provided. This change relaxes an existing validation.

fixes https://github.com/hashicorp/terraform-provider-google/issues/15064

I was unable to run the relevant acceptance tests because I do not have the permissions to create a network or subnetworks in the project owned by my employer. I built the provider locally and verified manually that I could create a GKE cluster with the expected configuration.

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [ ] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
container: 'master_ipv4_cidr_block' is not required when 'private_endpoint_subnetwork' is provided for 'google_container_cluster`
```

### CODE REVIEW

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 1]
```diff
@@ -4557,6 +4557,83 @@ resource "google_container_cluster" "with_private_endpoint_subnetwork" {
 `, containerNetName, s1Name, s1Cidr, s2Name, s2Cidr, clusterName)
 }
 
+func TestAccContainerCluster_withPrivateClusterConfigPrivateEndpointSubnetwork(t *testing.T) {
+	t.Parallel()
+
+	r := acctest.RandString(t, 10)
+
+	subnet1Name := fmt.Sprintf("tf-test-container-subnetwork1-%s", r)
+	subnet1Cidr := "10.0.36.0/24"
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4566]: Can we bake this & subnet2Cidr into the template? Values that don't need a random suffix or to be changed in an update test generally should be.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 2]
```diff
@@ -4557,6 +4557,83 @@ resource "google_container_cluster" "with_private_endpoint_subnetwork" {
...
+		CheckDestroy: testAccCheckContainerClusterDestroyProducer(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccContainerCluster_withPrivateClusterConfigPrivateEndpointSubnetwork(containerNetName, clusterName, subnet1Name, subnet1Cidr, subnet2Name, subnet2Cidr),
+			},
+			{
+				ResourceName:				"google_container_cluster.with_private_cluster_private_endpoint_subnetwork",
+				ImportState:				true,
+				ImportStateVerify:			true,
+				ImportStateVerifyIgnore:	[]string{"min_master_version"},
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4586]:
```suggestion
				ResourceName:            "google_container_cluster.with_private_cluster_private_endpoint_subnetwork",
				ImportState:             true,
				ImportStateVerify:       true,
				ImportStateVerifyIgnore: []string{"min_master_version"},
```

(can't tell if I got the spacing 100% right, but changed to spaces > tabs and aligned)

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 3]
```diff
@@ -4557,6 +4557,83 @@ resource "google_container_cluster" "with_private_endpoint_subnetwork" {
...
+  min_master_version = data.google_container_engine_versions.uscentral1a.release_channel_latest_version["STABLE"]
+  initial_node_count = 1
+
+  network    = google_compute_network.container_network.name
+  subnetwork = google_compute_subnetwork.container_subnetwork1.name
+
+  private_cluster_config {
+	enable_private_nodes    = true
+	enable_private_endpoint = true
+    private_endpoint_subnetwork = google_compute_subnetwork.container_subnetwork2.name
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4631]:
```suggestion
	enable_private_nodes    = true
	enable_private_endpoint = true
	private_endpoint_subnetwork = google_compute_subnetwork.container_subnetwork2.name
```

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 4]
```diff
@@ -4557,6 +4557,83 @@ resource "google_container_cluster" "with_private_endpoint_subnetwork" {
...
+  network                  = google_compute_network.container_network.name
+  ip_cidr_range            = "%s"
+  region                   = "us-central1"
+  private_ip_google_access = true
+}
+
+resource "google_container_cluster" "with_private_endpoint_subnetwork" {
+  name               = "%s"
+  location           = "us-central1-a"
+  min_master_version = data.google_container_engine_versions.uscentral1a.release_channel_latest_version["STABLE"]
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4622]: I think we can just omit this and we'll pick a reasonable version! that should remove the need to specify the datasource as well.

**Review by rileykarson — COMMENTED** [review]
Running tests. I can't determine whether this is an exhaustive change or not as the private cluster page doesn't mention private-endpoint-subnetwork, and the pages that do are about public clusters. The change seems reasonable, though.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb (hunk 1 above)
Comment (ncapps) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb]: I removed the `subnet1Cidr` and `subnet2Cidr` variables and function arguments. Please let me know if this is what you had in mind.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb (hunk 4 above)
Comment (rileykarson) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb]: That may cause a diff between the server-decided value and `latest`, I forget if we handle that. We can omit the field entirely to resolve if they come up.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb (hunk 1 above)
Comment (rileykarson) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb]: Yep!

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb (hunk 4 above)
Comment (ncapps) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb]: Sorry - I misunderstood your original comment. I removed the `min_master_version` field.

**Comment by rileykarson** [comment]
Looks like the test failed!

```
    vcr_utils.go:152: Step 1/2 error: Error running apply: exit status 1
        
        Error: googleapi: Error 400: Master authorized networks must be enabled if private endpoint is enabled.
        Details:
        [
          {
            "@type": "type.googleapis.com/google.rpc.RequestInfo",
            "requestId": "0xb13d7851caf9e50f"
          }
        ]
        , badRequest
        
          with google_container_cluster.with_private_endpoint_subnetwork,
          on terraform_plugin_test.tf line 23, in resource "google_container_cluster" "with_private_endpoint_subnetwork":
          23: resource "google_container_cluster" "with_private_endpoint_subnetwork" {
```

That means we'll need to configure https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/container_cluster#master_authorized_networks_config with some permutation that satisfies the validation. Several tests use `master_authorized_networks_config {}` so it's possible that's sufficient, but I'm not sure if this feature has harsher validaiton.

**Comment by ncapps** [comment]
Thank you for sharing the test result. I added the block below. This test is meant to represent a private cluster. Therefore I believe it is appropriate to disable access to the cluster master via GCE public IPs.

```tf
master_authorized_networks_config {
    gcp_public_cidrs_access_enabled = false
}
```

**Comment by ncapps** [comment]
Hi again,
I set the `networking_mode` to `VPC_NATIVE` to enable IP aliasing. This change requires the `ip_allocation_policy` block to be defined.

**Comment by rileykarson** [comment]
Running now!

**Comment by ncapps** [comment]
> 

Hi @rileykarson, Would you please send the latest test error message? I cannot access the logs details. Thank you!

**Comment by rileykarson** [comment]
```
    vcr_utils.go:152: Step 1/2 error: Error running apply: exit status 1
        
        Error: googleapi: Error 400: PrivateEndpointSubnetwork is not supported for VPC_PEERING private clusters.
        Details:
        [
          {
            "@type": "type.googleapis.com/google.rpc.RequestInfo",
            "requestId": "0xcba67b4f52feb76a"
          }
        ]
        , badRequest
        
          with google_container_cluster.with_private_endpoint_subnetwork,
          on terraform_plugin_test.tf line 25, in resource "google_container_cluster" "with_private_endpoint_subnetwork":
          25: resource "google_container_cluster" "with_private_endpoint_subnetwork" {
```

**Comment by ncapps** [comment]
I sent the http log of a successful gcloud request to your Google email account. Please let me know if I can provide additional information.

For what it's worth, I was able to create a cluster using Terraform with this configuration by building the terraform provider locally. I have yet to identify what is different between this test and my local build.

**Comment by rileykarson** [comment]
Thanks! I've received it.

There may be something weird with our test environment- it's added to a bunch of feature flags and allowlists. I'll push on the product team to figure out what's up, given it's working for you.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 5]
```diff
@@ -4597,6 +4597,82 @@ resource "google_container_cluster" "with_private_endpoint_subnetwork" {
...
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:	  func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		CheckDestroy: testAccCheckContainerClusterDestroyProducer(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccContainerCluster_withPrivateClusterConfigPrivateEndpointSubnetwork(containerNetName, clusterName),
+			},
+			{
+				ResourceName:            "google_container_cluster.with_private_cluster_private_endpoint_subnetwork",
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4617]:
```suggestion
				ResourceName:            "google_container_cluster.with_private_endpoint_subnetwork",
```

**Comment by rileykarson** [comment]
@ncapps I made & applied a suggestion to save a round trip, since it was a trivial change. Just a note that if we need to iterate further, you'll need to pull the change in from the origin branch yourself!

**Comment by ncapps** [comment]
Hi @rileykarson, Looks like the latest test run passed. Thanks so much for following up. Is there anything else I should do to help get this change ready to merge?

**Review by rileykarson — APPROVED** [review]
Nope, LGTM

**Pull request merged**